In [34]:
import pandas as pd
import numpy as np

In [35]:
data = pd.read_csv("./users.data", sep ="\t",names=["user_id", "item_id","rating","timestamp"])
data

user_id  item_id  rating  timestamp
0             0       50       5  881250949
1             0      172       5  881250949
2             0      133       1  881250949
3           196      242       3  881250949
4           186      302       3  891717742
...         ...      ...     ...        ...
99998       880      476       3  880175444
99999       716      204       5  879795543
100000      276     1090       1  874795795
100001       13      225       2  882399156
100002       12      203       3  879959583

[100003 rows x 4 columns]

### **VERİ OKUMA İŞLEMİNDE DÜZENLEME YAPILMAZSA VERİ BU ŞEKİLDE OLUR**

-	0\t50\t5\t881250949
- 0	0\t172\t5\t881250949
- 1	0\t133\t1\t881250949
- 2	196\t242\t3\t881250949
- 3	186\t302\t3\t891717742
- 4	22\t377\t1\t878887116
- ...	...
- 99997	880\t476\t3\t880175444
- 99998	716\t204\t5\t879795543
- 99999	276\t1090\t1\t874795795
- 100000	13\t225\t2\t882399156
- 100001	12\t203\t3\t879959583
- 100002 rows × 1 columns

In [36]:
data.isnull().sum()

user_id      0
item_id      0
rating       0
timestamp    0
dtype: int64

In [37]:
movies = pd.read_csv("./movie_id_titles.csv")
movies

item_id                                      title
0           1                           Toy Story (1995)
1           2                           GoldenEye (1995)
2           3                          Four Rooms (1995)
3           4                          Get Shorty (1995)
4           5                             Copycat (1995)
...       ...                                        ...
1677     1678                          Mat' i syn (1997)
1678     1679                           B. Monkey (1998)
1679     1680                       Sliding Doors (1998)
1680     1681                        You So Crazy (1994)
1681     1682  Scream of Stone (Schrei aus Stein) (1991)

[1682 rows x 2 columns]

In [38]:
movies.isnull().sum()

item_id    0
title      0
dtype: int64

In [39]:
# movies DataFrame'in title sütununu, data DataFrame'ine eklemek için merge işlemi yapıyoruz
data = pd.merge(data, movies[['item_id', 'title']], on="item_id")

# Yeni data DataFrame'inin ilk birkaç satırını görmek için
data.head()


user_id  item_id  rating  timestamp                            title
0        0       50       5  881250949                 Star Wars (1977)
1        0      172       5  881250949  Empire Strikes Back, The (1980)
2        0      133       1  881250949        Gone with the Wind (1939)
3      196      242       3  881250949                     Kolya (1996)
4      186      302       3  891717742         L.A. Confidential (1997)

In [40]:
# Yeni DataFrame'i CSV dosyasına yazdırıyoruz
data.to_csv("merge_data.csv", index=False)
data = pd.read_csv("./merge_data.csv", sep = ",")
data.head()

user_id  item_id  rating  timestamp                            title
0        0       50       5  881250949                 Star Wars (1977)
1        0      172       5  881250949  Empire Strikes Back, The (1980)
2        0      133       1  881250949        Gone with the Wind (1939)
3      196      242       3  881250949                     Kolya (1996)
4      186      302       3  891717742         L.A. Confidential (1997)

In [41]:
print(data[data["title"] == "Star Wars (1977)"]) # sadece star wars filmine verilen oyları görüntüledik

       user_id  item_id  rating  timestamp             title
0            0       50       5  881250949  Star Wars (1977)
502        290       50       5  880473582  Star Wars (1977)
860         79       50       4  891271545  Star Wars (1977)
1055         2       50       5  888552084  Star Wars (1977)
1093         8       50       5  879362124  Star Wars (1977)
...        ...      ...     ...        ...               ...
99570      749       50       5  878846978  Star Wars (1977)
99646      886       50       5  876031501  Star Wars (1977)
99856      739       50       4  886958895  Star Wars (1977)
99931      903       50       5  891031329  Star Wars (1977)
99951      780       50       5  891363685  Star Wars (1977)

[584 rows x 5 columns]


In [42]:
data.nunique() #kaç farklı veri olduğunu kontrol ediyoruz

user_id        944
item_id       1682
rating           5
timestamp    49282
title         1664
dtype: int64

In [43]:
table = data.pivot_table(index = "user_id", columns = "title", values = "rating" )
table.head()

title    'Til There Was You (1997)  1-900 (1994)  101 Dalmatians (1996)  \
user_id                                                                   
0                              NaN           NaN                    NaN   
1                              NaN           NaN                    2.0   
2                              NaN           NaN                    NaN   
3                              NaN           NaN                    NaN   
4                              NaN           NaN                    NaN   

title    12 Angry Men (1957)  187 (1997)  2 Days in the Valley (1996)  \
user_id                                                                 
0                        NaN         NaN                          NaN   
1                        5.0         NaN                          NaN   
2                        NaN         NaN                          NaN   
3                        NaN         2.0                          NaN   
4                        NaN         NaN                          NaN   

title    20,000 Leagues Under the Sea (1954)  2001: A Space Odyssey (1968)  \
user_id                                                                      
0                                        NaN                           NaN   
1                                        3.0                           4.0   
2                                        NaN                           NaN   
3                                        NaN                           NaN   
4                                        NaN                           NaN   

title    3 Ninjas: High Noon At Mega Mountain (1998)  39 Steps, The (1935)  \
user_id                                                                      
0                                                NaN                   NaN   
1                                                NaN                   NaN   
2                                                1.0                   NaN   
3                                                NaN                   NaN   
4                                                NaN                   NaN   

title    ...  Yankee Zulu (1994)  Year of the Horse (1997)  \
user_id  ...                                                 
0        ...                 NaN                       NaN   
1        ...                 NaN                       NaN   
2        ...                 NaN                       NaN   
3        ...                 NaN                       NaN   
4        ...                 NaN                       NaN   

title    You So Crazy (1994)  Young Frankenstein (1974)  Young Guns (1988)  \
user_id                                                                      
0                        NaN                        NaN                NaN   
1                        NaN                        5.0                3.0   
2                        NaN                        NaN                NaN   
3                        NaN                        NaN                NaN   
4                        NaN                        NaN                NaN   

title    Young Guns II (1990)  Young Poisoner's Handbook, The (1995)  \
user_id                                                                
0                         NaN                                    NaN   
1                         NaN                                    NaN   
2                         NaN                                    NaN   
3                         NaN                                    NaN   
4                         NaN                                    NaN   

title    Zeus and Roxanne (1997)  unknown  Á köldum klaka (Cold Fever) (1994)  
user_id                                                                        
0                            NaN      NaN                                 NaN  
1                            NaN      4.0                                 NaN  
2                            NaN      NaN                    

In [44]:
starwars = table["Star Wars (1977)"]
type(starwars)

pandas.core.series.Series

In [45]:
korelasyon = table.corrwith(starwars)
korelasyon

c:\Users\emirh\AppData\Local\Programs\Python\Python39\lib\site-packages\numpy\lib\function_base.py:2897: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
c:\Users\emirh\AppData\Local\Programs\Python\Python39\lib\site-packages\numpy\lib\function_base.py:2898: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
c:\Users\emirh\AppData\Local\Programs\Python\Python39\lib\site-packages\numpy\lib\function_base.py:2889: RuntimeWarning: Degrees of freedom <= 0 for slice
  c = cov(x, y, rowvar, dtype=dtype)
c:\Users\emirh\AppData\Local\Programs\Python\Python39\lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning: divide by zero encountered in divide
  c *= np.true_divide(1, fact)
c:\Users\emirh\AppData\Local\Programs\Python\Python39\lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning: invalid value encountered in multiply
  c *= np.true_divide(1, fact)


title
'Til There Was You (1997)                0.872872
1-900 (1994)                            -0.645497
101 Dalmatians (1996)                    0.211132
12 Angry Men (1957)                      0.184289
187 (1997)                               0.027398
                                           ...   
Young Guns II (1990)                     0.228615
Young Poisoner's Handbook, The (1995)   -0.007374
Zeus and Roxanne (1997)                  0.818182
unknown                                  0.723123
Á köldum klaka (Cold Fever) (1994)            NaN
Length: 1664, dtype: float64

In [46]:
yeni_korelasyon = pd.DataFrame(korelasyon, columns=["Correlation"])
yeni_korelasyon.dropna(inplace = True)
yeni_korelasyon.head()

Correlation
title                                 
'Til There Was You (1997)     0.872872
1-900 (1994)                 -0.645497
101 Dalmatians (1996)         0.211132
12 Angry Men (1957)           0.184289
187 (1997)                    0.027398

In [47]:
yeni_korelasyon.sort_values("Correlation", ascending = False).head(15)

Correlation
title                                                          
Hollow Reed (1996)                                          1.0
Commandments (1997)                                         1.0
Cosi (1996)                                                 1.0
No Escape (1994)                                            1.0
Stripes (1981)                                              1.0
Star Wars (1977)                                            1.0
Man of the Year (1995)                                      1.0
Beans of Egypt, Maine, The (1994)                           1.0
Old Lady Who Walked in the Sea, The (Vieille qu...          1.0
Outlaw, The (1943)                                          1.0
Line King: Al Hirschfeld, The (1996)                        1.0
Hurricane Streets (1998)                                    1.0
Scarlet Letter, The (1926)                                  1.0
Good Man in Africa, A (1994)                                1.0
Safe Passage (1994)                                         1.0

In [48]:
data.drop((["timestamp"]), axis = 1)

user_id  item_id  rating                            title
0             0       50       5                 Star Wars (1977)
1             0      172       5  Empire Strikes Back, The (1980)
2             0      133       1        Gone with the Wind (1939)
3           196      242       3                     Kolya (1996)
4           186      302       3         L.A. Confidential (1997)
...         ...      ...     ...                              ...
99998       880      476       3     First Wives Club, The (1996)
99999       716      204       5        Back to the Future (1985)
100000      276     1090       1                    Sliver (1993)
100001       13      225       2            101 Dalmatians (1996)
100002       12      203       3                Unforgiven (1992)

[100003 rows x 4 columns]

In [49]:
ratings = pd.DataFrame(data.groupby("title")["rating"].mean())
ratings.sort_values("rating", ascending= False).head()

rating
title                                             
They Made Me a Criminal (1939)                 5.0
Marlene Dietrich: Shadow and Light (1996)      5.0
Saint of Fort Washington, The (1993)           5.0
Someone Else's America (1995)                  5.0
Star Kid (1997)                                5.0

In [50]:
ratings["oy_sayısı"] = pd.DataFrame(data.groupby("title")["rating"].count())
ratings.head()

rating  oy_sayısı
title                                         
'Til There Was You (1997)  2.333333          9
1-900 (1994)               2.600000          5
101 Dalmatians (1996)      2.908257        109
12 Angry Men (1957)        4.344000        125
187 (1997)                 3.024390         41

In [51]:
ratings.sort_values("oy_sayısı", ascending = False ).head()

rating  oy_sayısı
title                                         
Star Wars (1977)           4.359589        584
Contact (1997)             3.803536        509
Fargo (1996)               4.155512        508
Return of the Jedi (1983)  4.007890        507
Liar Liar (1997)           3.156701        485

In [52]:
yeni_korelasyon = yeni_korelasyon.join(ratings ["oy_sayısı"])
yeni_korelasyon

Correlation  oy_sayısı
title                                                        
'Til There Was You (1997)                 0.872872          9
1-900 (1994)                             -0.645497          5
101 Dalmatians (1996)                     0.211132        109
12 Angry Men (1957)                       0.184289        125
187 (1997)                                0.027398         41
...                                            ...        ...
Young Guns (1988)                         0.186377        101
Young Guns II (1990)                      0.228615         44
Young Poisoner's Handbook, The (1995)    -0.007374         41
Zeus and Roxanne (1997)                   0.818182          6
unknown                                   0.723123          9

[1410 rows x 2 columns]

In [53]:
yeni_korelasyon[yeni_korelasyon ["oy_sayısı"] > 100].sort_values("Correlation", ascending = False).head()

Correlation  oy_sayısı
title                                                                     
Star Wars (1977)                                       1.000000        584
Empire Strikes Back, The (1980)                        0.748353        368
Return of the Jedi (1983)                              0.672556        507
Raiders of the Lost Ark (1981)                         0.536117        420
Austin Powers: International Man of Mystery (1997)     0.377433        130